src/rsz/src/Rebuffer.cc

```diff
@@ -227,55 +240,44 @@ RepairSetup::rebufferBottomUp(const BufferedNetPtr& bnet,
         junc->setCapacitance(p->cap() + q->cap());
         junc->setRequiredPath(min_req->requiredPath());
         junc->setRequiredDelay(min_req->requiredDelay());
-        Z.push_back(std::move(junc));
+        slacks[junc.get()] = slackPenalized(junc);
+        Z[base + j] = std::move(junc);
       }
     }
     // Prune the options if there exists another option with
     // larger required and smaller capacitance.
-    // This is fanout^2.
-    // Presort options to hit better options sooner.
+    // This is fanout*log(fanout) if options are
+    // presorted to hit better options sooner.
     sort(Z.begin(),
          Z.end(),
-         [this](const BufferedNetPtr& option1, const BufferedNetPtr& option2) {
-           Slack slack1 = slackPenalized(option1);
-           Slack slack2 = slackPenalized(option2);
+         [&slacks](const BufferedNetPtr& option1, const BufferedNetPtr& option2) {
+           Slack slack1 = slacks[option1.get()];
+           Slack slack2 = slacks[option2.get()];
 
-           if (slack1 > slack2) {
-             return true;
-           }
-           if (slack2 > slack1) {
-             return false;
+           if (slack1 != slack2) {
+             return slack1 > slack2;
            }
 
-           if (option1->cap() < option2->cap()) {
-             return true;
-           }
-           if (option2->cap() < option1->cap()) {
-             return false;
-           }
-
-           return false;
+           return option1->cap() < option2->cap();
          });
-    int si = 0;
-    for (size_t pi = 0; pi < Z.size(); pi++) {
+    float Lsmall = Z[0]->cap();
+    size_t si = 1;
+    // Remove options by shifting down with index si.
+    // Because the options are sorted we don't have to look
+    // beyond the first option. We also know that slack
+    // is nonincreasing, so we can remove everything that has
+    // higher capacitance than the lowest found so far.
+    for (size_t pi = si; pi < size; pi++) {
       const BufferedNetPtr& p = Z[pi];
       float Lp = p->cap();
-      // Remove options by shifting down with index si.
-      si = pi + 1;
-      // Because the options are sorted we don't have to look
-      // beyond the first option.
-      for (size_t qi = pi + 1; qi < Z.size(); qi++) {
-        const BufferedNetPtr& q = Z[qi];
-        float Lq = q->cap();
-        // We know Tq <= Tp from the sort so we don't need to check req.
-        // If q is the same or worse than p, remove solution q.
-        if (fuzzyLess(Lq, Lp)) {
-          // Copy survivor down.
-          Z[si++] = q;
-        }
+      // If Tp is the same or worse than Tsmall, remove solution p.
+      if (fuzzyLess(Lp, Lsmall)) {
+        // Otherwise copy the survivor down.
+        Z[si++] = p;
+        Lsmall = Lp;
```

Comment:
The previous code did not update Lp so it likely kept many more options.  This is a change in functionality not a pure performance tuning and will require validation if is intended - please confirm.   It sounds reasonable.  I expect this is where most of the performance gain comes from.

Reply:
Yes, it is intended. I changed it because an improvement of complexity seemed reasonable, but in the designs I tested most of improvement came not from this, but from caching the slack values (and so avoiding redundant floating point calculations).

Reply:
The previous code would iterate over the vector and for each element it'd remove its successors with a higher or similar capacitance. If some element is not removed, it must have a lower capacitance, so on the next iteration the inner loop will iterate over the survivors again and challenge them even more. It should be enough to iterate once, while keeping track of the lowest cap so far.

I think If this argument is correct, then both versions should produce the same output.